### ISSUE TITLE
design/write general automation process for pushing documentation

### ISSUE THREAD

**Issue by jberkus** [projectatomic/docs.projectatomic.io#5]
Figure out how all of our constituent projects are going to push documentation to the new gear, and then implement the tooling.  Will lead to additional tasks.

@mscherer @jlebon @bexelbie

**Comment by bproffitt** [comment]
Also, coordinate with @smccance? (Suggestion)

Brian Proffitt
@TheTechScribe

On Mar 22, 2016, at 7:19 PM, Josh Berkus <notifications@github.com> wrote:

Figure out how all of our constituent projects are going to push
documentation to the new gear, and then implement the tooling. Will lead to
additional tasks.

@mscherer <https://github.com/mscherer> @jlebon <https://github.com/jlebon>
@bexelbie <https://github.com/bexelbie>

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub
<https://github.com/projectatomic/docs.projectatomic.io/issues/5>

**Comment by jberkus** [comment]
@bproffitt not sure who that is or why?

**Comment by mscherer** [comment]
You mean @shaunix (Shaun Mc Cance) ?

**Comment by bproffitt** [comment]
Yep, my mistake

Brian Proffitt
@TheTechScribe

On Mar 23, 2016, at 3:49 AM, Michael Scherer <notifications@github.com>
wrote:

You mean @shaunix <https://github.com/shaunix> (Shaun Mc Cance) ?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/projectatomic/docs.projectatomic.io/issues/5#issuecomment-200233369>

**Comment by jlebon** [comment]
> Figure out how all of our constituent projects are going to push documentation to the new gear, and then implement the tooling. Will lead to additional tasks.

I guess at a high level:

- Each individual project provides a mechanism for building their documentation, preferably using docker containers. The resulting HTML files should be compatible with our (future) stylesheet.
- We set up an automated job (hosted somewhere TBD) defined by files maintained in this repo in which:
  - For each project (i.e. all the upstream doc repos + this repo, which will host some docs as well), check whether docs have been updated.
  - For each updated project, run the build procedure (e.g. `docker run`) to produce final artifacts.
  - Rsync all produced artifacts to gear.

Am I missing something?

If we decide to use Jenkins, I can help writing the job definition in YAML (using JJB). This is a format that is fairly approachable, even for folks unfamiliar with Jenkins. Actually hosting the job is a different issue. If it's not required for it to be accessible publicly, we could probably use the internal atomic-e2e Jenkins (this would not preclude the ability to e.g. post all the logs/results from builds to PRs etc..., it would only mean that the general public wouldn't have direct access to the Jenkins interface).

**Comment by jberkus** [comment]
We're project Atomic, we can *require* a container for doc build.  

Question is, how do we get notified of per-project doc updates so that we can build and push those docs?

I'm fine with the public not having access to the Jenkins interface for now.  We can revisit that when it becomes an inhibitor to outside contribution, maybe in a year.

**Comment by jlebon** [comment]
> We're project Atomic, we can require a container for doc build. 

Agreed.

> Question is, how do we get notified of per-project doc updates so that we can build and push those docs?

This is not an issue in Jenkins (and I'm sure most automation tools). There is support for both polling and push notifications (using webhooks, though I highly doubt any of the docs projects will be that time critical). It would probably be better to have a very slow poll e.g. once or twice per day so that we can coalesce updates across all projects in a single rsync.

**Comment by jberkus** [comment]
I thought that several teams (Best Practices, Cockpit, etc.) wanted to have more rapid builds than once per day?

**Comment by jlebon** [comment]
Sure, this is something that should be customizable. E.g. default to slow poll, but fast poll/push for select projects.

**Comment by mscherer** [comment]
In a ideal world, using openshift v3 solve this. We can make a pod with all the required containers and it should do the right thing(tm).

**Comment by bproffitt** [comment]
Shaun McCance is the documentation specialist for OSAS and should be an
excellent resource for this project.

Brian Proffitt
@TheTechScribe

On Mar 22, 2016, at 8:15 PM, Josh Berkus <notifications@github.com> wrote:

@bproffitt <https://github.com/bproffitt> not sure who that is or why?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/projectatomic/docs.projectatomic.io/issues/5#issuecomment-200090202>

**Comment by jberkus** [comment]
@mscherer what's the timeline for v3 Online being available?

**Comment by jberkus** [comment]
@bproffitt I chatted with Shaun.  He did not have any suggestions.

**Comment by mscherer** [comment]
I do not have any date for v3, but I can ask to the ops team today to see if they have a answer, and go search the intranet.

**Comment by jberkus** [comment]
I don't think we can wait for v3.

Here's my proposal:

* Each project will supply a container (definition) for building documentation.
* The doc build container will accept a relative path as a run parameter.  It will use this relative path for all links inside the documentation if there are internal links.
* The container will output HTML documentation to a mapped volume, at /docs/
* The resulting HTML will link projectatomic styling, possibly by referencing projectatomic.css in the root.

In other words, the build system should be able to pull the project, build the docs container, and then use it to emit HTML documentation which can be synced to the OpenShift instance.

Alternately, we can ask projects to build and push a container image, but I'm not sure how we'd get notified when it was updated.
